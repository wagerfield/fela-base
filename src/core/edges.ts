export const ALL = ["top", "right", "bottom", "left"]
export const TRB = ["top", "right", "bottom"]
export const TLB = ["top", "left", "bottom"]
export const LBR = ["left", "bottom", "right"]
export const LTR = ["left", "top", "right"]
export const BR = ["bottom", "right"]
export const BL = ["bottom", "left"]
export const LR = ["left", "right"]
export const TB = ["top", "bottom"]
export const TR = ["top", "right"]
export const TL = ["top", "left"]
export const B = ["bottom"]
export const R = ["right"]
export const L = ["left"]
export const T = ["top"]
