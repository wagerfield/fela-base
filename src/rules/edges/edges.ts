import { arrayReduce, objectReduce } from "fast-loops"
import {
  isNull,
  isArray,
  isObject,
  parseEdges,
  wrapKey
} from "../../core/utils"

export interface EdgesStyle {
  [key: string]: any
}

export type Edges = string | string[] | EdgesStyle

export interface EdgesRuleOptions {
  edges?: Edges
  value?: any
  prefix?: string
  postfix?: string
}

export default ({
  value,
  edges,
  prefix,
  postfix
}: EdgesRuleOptions = {}) => {
  const keys = parseEdges(edges)
  const edgesStyle: EdgesStyle = {}
  // Handle null values
  if (isNull(value)) {
    return edgesStyle
  } else if (isArray(keys)) {
    // Process edge arrays
    return arrayReduce(
      keys,
      (style, key) => {
        style[wrapKey(key, prefix, postfix)] = value
        return style
      },
      edgesStyle
    )
  } else if (isObject(keys)) {
    // Process edge objects
    return objectReduce(
      keys,
      (style, keyValue, key) => {
        style[wrapKey(key, prefix, postfix)] = keyValue
        return style
      },
      edgesStyle
    )
  } else {
    throw new Error("invalid edges value")
  }
}
