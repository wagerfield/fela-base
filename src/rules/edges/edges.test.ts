import edges from "./edges"

describe("edges", () => {
  it("throws error with no edges", () => {
    expect(() => {
      edges()
    }).toThrowErrorMatchingSnapshot()
  })

  it("throws error with invalid edges", () => {
    expect(() => {
      edges({ edges: null })
    }).toThrowErrorMatchingSnapshot()
  })

  it("renders expected snapshot when value is null", () => {
    expect(
      edges({
        edges: "top",
        value: null
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edges is a string", () => {
    expect(
      edges({
        edges: "left|right",
        value: "foo"
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edges is an array", () => {
    expect(
      edges({
        edges: ["left", "right"],
        value: "foo"
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot when edges is an object", () => {
    expect(
      edges({
        edges: {
          left: "foo",
          right: "bar"
        }
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot with prefixed props", () => {
    expect(
      edges({
        value: "10px",
        edges: "left|right",
        prefix: "padding"
      })
    ).toMatchSnapshot()
  })

  it("renders expected snapshot with postfixed props", () => {
    expect(
      edges({
        value: "red",
        edges: "left|right",
        prefix: "border",
        postfix: "color"
      })
    ).toMatchSnapshot()
  })
})
