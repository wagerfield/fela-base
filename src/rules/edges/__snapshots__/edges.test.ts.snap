// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`edges renders expected snapshot when edges is a string 1`] = `
Object {
  "left": "foo",
  "right": "foo",
}
`;

exports[`edges renders expected snapshot when edges is an array 1`] = `
Object {
  "left": "foo",
  "right": "foo",
}
`;

exports[`edges renders expected snapshot when edges is an object 1`] = `
Object {
  "left": "foo",
  "right": "bar",
}
`;

exports[`edges renders expected snapshot when value is null 1`] = `Object {}`;

exports[`edges renders expected snapshot with postfixed props 1`] = `
Object {
  "borderLeftColor": "red",
  "borderRightColor": "red",
}
`;

exports[`edges renders expected snapshot with prefixed props 1`] = `
Object {
  "paddingLeft": "10px",
  "paddingRight": "10px",
}
`;

exports[`edges throws error with invalid edges 1`] = `"invalid edges value"`;

exports[`edges throws error with no edges 1`] = `"invalid edges value"`;
